packages/@okta/vuepress-site/docs/guides/oie-embedded-sdk-use-case-custom-pwd-recovery-mfa/main/java/sdk.md

```diff
@@ -0,0 +1,110 @@
+### 1: Start password recovery
+
+The user starts the password recovery flow with the following steps:
+
+1. Clicks the **Forgot Password?** link on the sign-in page.
+2. Enters their **Email or Username** in the dialog and then clicks **Next**.
+3. Chooses **Email** as the authenticator they want to use for password recovery and clicks **Submit**.
+
+Okta then sends an email to the email address that matches the Forgot Password template that was altered earlier.
+
+<div class="common-image-format">
+
+![Example of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png "Password recovery email")
+
+</div>
+
+The email's **Reset Password** link includes the `otp` and `request.relayState` variables sent back as query parameters to the application. For instance, the URL in the email template,  `http://localhost:8080/magic-link/callback?otp=${oneTimePassword}&state=${request.relayState}`, might be rendered as `http://localhost:8080/magic-link/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32` in the email sent to the user.
+
+### 2: Handle the OTP and state parameters
+
+Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as parameters to the `IdxAuthenticationWrapper.verifyAuthenticator()` method. Before calling the method, check that the `otp` and `state` parameters have values. If either parameter is `null` or the value of `state` does not match the state stored in the current `ProceedContext` session variable, throw an error.
```

Comment:
It's unclear whether the handler takes a query string from which it parses the parameters or if it takes two parameters that are the result of parsing the query string. I think it's the former. It's also not quite in sequence as it the call to `verifyAuthenticator()` occurs before checking for `null`. Finally, isn't "callback handler method" redundant, that is, isnt' a "callback handler" a method? How about something like:

> Create a callback handler that takes the `otp` and `state` parameters from the query string. Check that neither value is `null` and that the value of `state` doesn't match the state stored in the current `ProceedContext` session variable. In either of these cases, thrown an error. Otherwise, call `IdxAuthenticationWrapper.verifyAuthenticator()` using ?WHAT? (there's multiple calls below, so which parameter is it using, if any)?

Reply:
Have changed to 

> Create a callback handler that takes the `otp` and `state` parameters from the query string, and checks they have values that are valid:
> 
> 1. Check `otp` and `state` are not `null`.
> 2. Check `state` matches the state stored in the current `ProceedContext` session variable.
> 
> If either check returns false, throw an error.

and just left out the end action of the handler until we get to it in the code

Reply:
This still isn't quite clear because it's unclear who'se calling the handler. It's also possible to interpret the current wording as "you as the developer need to parse the `otp` and state `parameters` for your handler, then call it." Where does `state` come from? And why "current `ProceedContext` session variable", is it possible to have multiple concurrent `ProceedContext` session variables?How about:

> Create a callback handler that's called <possibly WHEN> <by WHO>. The handler takes two parameters, the value of the `otp` and `state` parameters from the query string in the URL, and checks that:
> 
> 1. The parameters are non-`null`.
> 2. The value of `state` matches the value stored in the `ProceedContext` session variable.

Reply:
OK have consulted how it's explaied in the java docs and changed to

> Create a mapping that maps a request for the magic link URL to a method that takes the `otp` and `state` parameters from the query string, and makes the following checks:
> 
> 1. That `otp` and `state` are not `null`.
> 2. That `state` matches the state stored in the current `ProceedContext` session variable.
> 
> If either check returns false, throw an error.

Reply:
First of, I'm not seeing anything in the current version of the committed file. Second, unless I'm missing something that any Java developer knows, the action required by the developer still isn't clear (I suspect the Java version you looked at is poorly explained as well.)

Start from first principles. What is the task in this section? What are the coding steps required to handle that task? Where does that code reside and in what language (the wording above makes it sound like the "mapping" is in the HTML world and the rest is in Java)? Is this conflating multiple steps? For example, is there a step to create something that extracts the `otp` and `state` query string parameters (which is clearer than "parameters from the query string"), then another for calling a handler, and third for creating the handler? Who calls the handler: the code in this step, Okta, the host browser, ...?

Reply:
BTW, it's "receives", not "takes"